zoo_frontend/src/pages/admin/user/user.jsx

```diff
@@ -1,50 +1,396 @@
+
+class User extends Component {
   /**
    * Server side data retrieval
    */
   static async getInitialProps({ authToken }) {
-    const api = new UsersAPI(authToken);
-    const res = await api.getUsers().catch((err) => ({ data: [{ err: true, msg: err }] }));
-    return { users: res.data };
+    const usersApi = new UsersAPI(authToken);
+    const departmentsApi = new DepartmentAPI(authToken);
+    const rolesApi = new RolesAPI(authToken);
+    const roleMappings = new RoleMappingsAPI(authToken);
+    try {
+      const [usersRes, locRes, rolesRes, roleMapRes] = await Promise.all([
+        usersApi.getUsers(),
+        departmentsApi.getDepartments(),
+        rolesApi.getRoles(),
+        roleMappings.getRoleMappings(),
+      ]);
+      return {
+        users: usersRes.data,
+        locations: locRes.data,
+        roles: rolesRes.data,
+        roleMappings: roleMapRes.data,
+      };
+    } catch (err) {
+      return {
+        users: [],
+        locations: [],
+        roles: [],
+        roleMappings: [],
+        error: true,
+        errorMessage: 'Error loading data.',
+      };
+    }
   }
 
   static propTypes = {
-    account: PropTypes.object.isRequired,
+    // account: PropTypes.object.isRequired,
     classes: PropTypes.object.isRequired,
+    error: PropTypes.bool,
+    errorMessage: PropTypes.string,
+    locations: PropTypes.arrayOf(PropTypes.object).isRequired,
+    roleMappings: PropTypes.arrayOf(PropTypes.object).isRequired,
+    roles: PropTypes.arrayOf(PropTypes.object).isRequired,
     token: PropTypes.string,
-    users: PropTypes.array.isRequired,
+    users: PropTypes.arrayOf(PropTypes.object).isRequired,
   };
 
   static defaultProps = {
+    error: false,
+    errorMessage: '',
     token: '',
   }
 
   constructor(props) {
     super(props);
+
     this.state = {
-      asdf: props.token, // eslint-disable-line react/no-unused-state
+      token: props.token,
+      ...generateStateData(this.props.users, this.props.locations, this.props.roles, this.props.roleMappings),
+      locations: this.props.locations,
+      roles: this.props.roles,
+      passwordDialogOpen: false,
+      passwordDialogQuestion: '',
+      passwordText: '',
     };
+
+    this.passwordDialogResolve = null;
+
+    this.notificationsRef = React.createRef();
+  }
+
+  getPassword = (question) => new Promise((resolve) => {
+    this.setState({
+      passwordDialogOpen: true,
+      passwordDialogQuestion: question,
+      passwordText: 'OmahaZoo',
+    });
+    this.passwordDialogResolve = resolve;
+  });
+
+  onRowAdd = (newData) => new Promise(async (resolve, reject) => {
+    // Reject the new user if there is no firstname, lastname, email, role, or location
+    if (!newData.firstName || !newData.lastName || !newData.email || !newData.role || !newData.locationName) {
+      this.notificationsRef.current.showNotification('error', 'Please fill out all of the fields to create a new user.');
+      reject();
+      return;
```

Comment:
Is the return necessary if you have reject above it?

Reply:
Yes because resolve/reject doesn't not exit the function.